lib/internal/loader/ModuleRequest.js

```diff
@@ -36,24 +38,39 @@ loaders.set('esm', async (url) => {
 
 // Strategy for loading a node-style CommonJS module
 loaders.set('cjs', async (url) => {
-  return createDynamicModule(['default'], url, (reflect) => {
-    debug(`Loading CJSModule ${url}`);
-    const CJSModule = require('module');
-    const pathname = internalURLModule.getPathFromURL(new URL(url));
-    CJSModule._load(pathname);
+  debug(`Loading CJSModule ${url}`);
+  const CJSModule = require('module');
+  const pathname = internalURLModule.getPathFromURL(new URL(url));
+  const exports = CJSModule._load(pathname);
+  const es = exports[esModuleInterop] !== undefined ?
+    exports[esModuleInterop] : exports.__esModule;
+  const keys = es ? Object.keys(exports) : ['default'];
+  return createDynamicModule(keys, url, (reflect) => {
+    if (es) {
+      for (let i = 0; i < keys.length; i++)
+        reflect.exports[keys[i]].set(exports[keys[i]]);
+    } else {
+      reflect.exports.default.set(exports);
+    }
   });
 });
 
 // Strategy for loading a node builtin CommonJS module that isn't
 // through normal resolution
 loaders.set('builtin', async (url) => {
-  return createDynamicModule(['default'], url, (reflect) => {
-    debug(`Loading BuiltinModule ${url}`);
-    const exports = NativeModule.require(url.substr(5));
+  debug(`Loading BuiltinModule ${url}`);
+  const exports = NativeModule.require(url.substr(5));
+  const keys = Object.keys(exports);
+  return createDynamicModule(['default', ...keys], url, (reflect) => {
     reflect.exports.default.set(exports);
+    for (let i = 0; i < keys.length; i++)
+      reflect.exports[keys[i]].set(exports[keys[i]]);
   });
 });
 
+// Strategy for loading a native addon module
+// Named exports will not be parsed from these - see
+// https://github.com/nodejs/abi-stable-node/issues/256#issuecomment-325138872
```

Comment:
Fwiw, another thing I’d like to see (not necessarily in this PR) is keeping parity between existing addons and CJS modules. I agree with the linked comment in that we don’t need any extra wiring for ESM, at least for now, but I don’t see any good reason to let wrapping for CJS and native addons diverge.

Reply:
i read this as napi wanting to keep a single export point such that they don't need to worry about if its `default` or `module.exports` or whatever else it may be, and changing how that gets exposed on the node side would be kinda weird imo

Reply:
It’s certainly not that important for addons, but … if `require('foo')` and `require('foo.node')` do the same thing, then I think `import 'foo';` and `import 'foo.node';` should also do the same thing

Reply:
@addaleax the `foo.node` package can create an ESM module wrapper that sets up named imports.